File: tests/docker-compat/go-client/tests/transactions.go
package tests

import (
	"context"
	"fmt"

	"github.com/redis/go-redis/v9"
)

func init() {
	Register("transactions", testTransactions)
}

func testTransactions(ctx context.Context, rdb *redis.Client, prefix string) []TestResult {
	var results []TestResult
	cat := "transactions"

	results = append(results, RunTest("MULTI_EXEC_basic", cat, func() error {
		k1 := Key(prefix, "tx:basic1")
		k2 := Key(prefix, "tx:basic2")
		defer rdb.Del(ctx, k1, k2)

		pipe := rdb.TxPipeline()
		pipe.Set(ctx, k1, "a", 0)
		pipe.Set(ctx, k2, "b", 0)
		_, err := pipe.Exec(ctx)
		if err != nil {
			return err
		}

		v1, _ := rdb.Get(ctx, k1).Result()
		v2, _ := rdb.Get(ctx, k2).Result()
		if err := AssertEqual("a", v1); err != nil {
			return err
		}
		return AssertEqual("b", v2)
	}))

	results = append(results, RunTest("MULTI_DISCARD", cat, func() error {
		k := Key(prefix, "tx:discard")
		defer rdb.Del(ctx, k)
		rdb.Set(ctx, k, "original", 0)

		// Use raw commands for MULTI/DISCARD
		err := rdb.Do(ctx, "MULTI").Err()
		if err != nil {
			return err
		}
		rdb.Do(ctx, "SET", k, "changed")
		err = rdb.Do(ctx, "DISCARD").Err()
		if err != nil {
			return err
		}

		val, _ := rdb.Get(ctx, k).Result()
		return AssertEqual("original", val)
	}))

	results = append(results, RunTest("WATCH_UNWATCH_optimistic_lock", cat, func() error {
		k := Key(prefix, "tx:watch")
		defer rdb.Del(ctx, k)
		rdb.Set(ctx, k, "10", 0)

		// WATCH + no modification = EXEC should succeed
		err := rdb.Watch(ctx, func(tx *redis.Tx) error {
			val, err := tx.Get(ctx, k).Result()
			if err != nil {
				return err
			}
			if val != "10" {
				return fmt.Errorf("unexpected val: %s", val)
			}
			_, err = tx.TxPipelined(ctx, func(pipe redis.Pipeliner) error {
				pipe.Set(ctx, k, "20", 0)
				return nil
			})
			return err
		}, k)
		if err != nil {
			return err
		}

		val, _ := rdb.Get(ctx, k).Result()
		return AssertEqual("20", val)
	}))

	results = append(results, RunTest("WATCH_key_modification_detection", cat, func() error {
		k := Key(prefix, "tx:watchmod")
		defer rdb.Del(ctx, k)
		rdb.Set(ctx, k, "original", 0)

		// Use a second client to modify the key between WATCH and EXEC
		rdb2 := redis.NewClient(&redis.Options{
			Addr:     rdb.Options().Addr,
			Password: rdb.Options().Password,
		})
		defer rdb2.Close()

		err := rdb.Watch(ctx, func(tx *redis.Tx) error {
			// Another client modifies the watched key
			rdb2.Set(ctx, k, "modified_by_other", 0)

			_, err := tx.TxPipelined(ctx, func(pipe redis.Pipeliner) error {
				pipe.Set(ctx, k, "should_fail", 0)
				return nil
			})
			return err
		}, k)

		// Should get TxFailedErr because the key was modified
		if err != redis.TxFailedErr {
			return fmt.Errorf("expected TxFailedErr, got: %v", err)
		}

		val, _ := rdb.Get(ctx, k).Result()
		return AssertEqual("modified_by_other", val)
	}))

	results = append(results, RunTest("queued_command_errors", cat, func() error {
		k := Key(prefix, "tx:qerr")
		defer rdb.Del(ctx, k)
		rdb.Set(ctx, k, "string_value", 0)

		pipe := rdb.TxPipeline()
		set := pipe.Set(ctx, k, "new_value", 0)
		// This will fail because k is a string, not a list
		lpush := pipe.LPush(ctx, k, "bad")
		_, err := pipe.Exec(ctx)
		// Exec returns error but individual commands may succeed
		_ = err

		// The SET should have succeeded
		if set.Err() != nil {
			return fmt.Errorf("SET in tx should succeed: %v", set.Err())
		}
		// The LPUSH should have failed with WRONGTYPE
		if lpush.Err() == nil {
			return fmt.Errorf("LPUSH should fail with WRONGTYPE")
		}
		return nil
	}))

	return results
}

File: tests/docker-compat/js-client/tests/transactions.js
const Redis = require('ioredis');
const { register, runTest, key, assertEqual, assertTrue } = require('./framework');

register('transactions', async (redis, prefix) => {
  const results = [];
  const cat = 'transactions';

  results.push(await runTest('MULTI_EXEC_basic', cat, async () => {
    const k1 = key(prefix, 'tx:basic1');
    const k2 = key(prefix, 'tx:basic2');
    const result = await redis.multi()
      .set(k1, 'a')
      .set(k2, 'b')
      .exec();
    // exec returns [[null, 'OK'], [null, 'OK']]
    assertTrue(result !== null, 'EXEC should return results');
    const v1 = await redis.get(k1);
    const v2 = await redis.get(k2);
    assertEqual('a', v1);
    assertEqual('b', v2);
    await redis.del(k1, k2);
  }));

  results.push(await runTest('MULTI_DISCARD', cat, async () => {
    const k = key(prefix, 'tx:discard');
    await redis.set(k, 'original');
    // Use a dedicated connection to ensure MULTI/DISCARD run on the same connection
    const conn = redis.duplicate();
    try {
      await conn.connect();
      await conn.call('MULTI');
      await conn.call('SET', k, 'changed');
      await conn.call('DISCARD');
    } finally {
      conn.disconnect();
    }
    const val = await redis.get(k);
    assertEqual('original', val);
    await redis.del(k);
  }));

  results.push(await runTest('WATCH_UNWATCH_optimistic_lock', cat, async () => {
    const k = key(prefix, 'tx:watch');
    await redis.set(k, '10');

    // WATCH + no modification = EXEC should succeed
    await redis.watch(k);
    const val = await redis.get(k);
    assertEqual('10', val);
    const result = await redis.multi()
      .set(k, '20')
      .exec();
    assertTrue(result !== null, 'EXEC should succeed when watched key not modified');
    const newVal = await redis.get(k);
    assertEqual('20', newVal);
    await redis.del(k);
  }));

  results.push(await runTest('WATCH_key_modification_detection', cat, async () => {
    const k = key(prefix, 'tx:watchmod');
    await redis.set(k, 'original');

    const rdb2 = new Redis({
      host: redis.options.host,
      port: redis.options.port,
      password: redis.options.password,
      lazyConnect: true,
    });
    await rdb2.connect();

    try {
      await redis.watch(k);
      // Another client modifies the watched key
      await rdb2.set(k, 'modified_by_other');
      const result = await redis.multi()
        .set(k, 'should_fail')
        .exec();
      // When WATCH detects modification, exec() returns null
      assertTrue(result === null, 'EXEC should return null when watched key was modified');
      const val = await redis.get(k);
      assertEqual('modified_by_other', val);
    } finally {
      rdb2.disconnect();
      await redis.del(k);
    }
  }));

  results.push(await runTest('queued_command_errors', cat, async () => {
    const k = key(prefix, 'tx:qerr');
    await redis.set(k, 'string_value');
    const result = await redis.multi()
      .set(k, 'new_value')
      .lpush(k, 'bad')  // This will fail because k is a string
      .exec();
    // In ioredis, exec returns array with [err, result] for each command
    assertTrue(result !== null, 'EXEC should return results');
    assertTrue(result[0][0] === null, 'SET should succeed');
    assertTrue(result[1][0] !== null, 'LPUSH should fail with WRONGTYPE');
    await redis.del(k);
  }));

  return results;
});
